File: libs/query-engine/src/flight_provider.rs
//! Apache Arrow Flight SQL [`TableProvider`] for DataFusion.
//!
//! [`FlightSqlTableProvider`] turns a remote Flight SQL endpoint into a
//! regular DataFusion table. Any data-plane service can therefore federate
//! queries against another Flight SQL service exactly as if its result set
//! were a local table:
//!
//! ```ignore
//! use std::sync::Arc;
//! use datafusion::prelude::SessionContext;
//! use query_engine::flight_provider::FlightSqlTableProvider;
//!
//! # async fn run() -> datafusion::error::Result<()> {
//! let provider = FlightSqlTableProvider::try_new(
//!     "http://flight-sql.internal:50051",
//!     "SELECT id, name FROM customers",
//! )
//! .await
//! .unwrap();
//!
//! let ctx = SessionContext::new();
//! ctx.register_table("customers", Arc::new(provider))?;
//! let df = ctx.sql("SELECT * FROM customers LIMIT 10").await?;
//! df.show().await?;
//! # Ok(()) }
//! ```
//!
//! ## Push-down support
//!
//! * **Projection** is pushed down trivially: only the requested columns
//!   are kept once the batches arrive from the remote endpoint.
//! * **Limit** is pushed down trivially: the local stream stops as soon as
//!   `limit` rows have been emitted.
//! * **Filter** push-down is intentionally **out of scope**. Translating
//!   arbitrary DataFusion `Expr`s into the SQL dialect of an unknown remote
//!   engine cannot be done safely in a generic provider. Callers that need
//!   filter push-down should embed the predicate directly in the `query`
//!   they pass to [`FlightSqlTableProvider::try_new`].

use std::any::Any;
use std::fmt;
use std::sync::Arc;

use arrow::datatypes::SchemaRef;
use arrow::record_batch::RecordBatch;
use arrow_flight::sql::client::FlightSqlServiceClient;
use async_trait::async_trait;
use datafusion::catalog::Session;
use datafusion::common::{DataFusionError, project_schema};
use datafusion::datasource::{TableProvider, TableType};
use datafusion::error::Result as DfResult;
use datafusion::execution::{SendableRecordBatchStream, TaskContext};
use datafusion::logical_expr::{Expr, TableProviderFilterPushDown};
use datafusion::physical_expr::EquivalenceProperties;
use datafusion::physical_plan::execution_plan::{Boundedness, EmissionType};
use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
use datafusion::physical_plan::{
    DisplayAs, DisplayFormatType, ExecutionPlan, Partitioning, PlanProperties,
};
use futures::{StreamExt, TryStreamExt, stream};
use thiserror::Error;
use tonic::transport::{Channel, Endpoint};

/// Errors raised while building or driving a [`FlightSqlTableProvider`].
#[derive(Debug, Error)]
pub enum FlightProviderError {
    /// The supplied endpoint URL could not be parsed or reached.
    #[error("invalid Flight SQL endpoint `{endpoint}`: {source}")]
    InvalidEndpoint {
        endpoint: String,
        #[source]
        source: tonic::transport::Error,
    },
    /// The TCP/HTTP-2 connection to the remote Flight server failed.
    #[error("failed to connect to Flight SQL endpoint `{endpoint}`: {source}")]
    Connect {
        endpoint: String,
        #[source]
        source: tonic::transport::Error,
    },
    /// The remote Flight SQL server returned an Arrow-level error.
    #[error("Flight SQL error: {0}")]
    Flight(String),
    /// The Flight server replied with a [`FlightInfo`] that did not contain
    /// any endpoint (and therefore no ticket) for the query.
    #[error("Flight SQL endpoint returned no ticket for query")]
    MissingTicket,
}

impl From<FlightProviderError> for DataFusionError {
    fn from(value: FlightProviderError) -> Self {
        DataFusionError::External(Box::new(value))
    }
}

/// A [`TableProvider`] that resolves its data by issuing a SQL query against
/// a remote Apache Arrow Flight SQL endpoint.
///
/// The schema is supplied up-front by the caller because DataFusion needs it
/// at planning time, before any I/O takes place. The remote service must
/// produce a result-set that is compatible with this schema.
#[derive(Debug, Clone)]
pub struct FlightSqlTableProvider {
    endpoint: String,
    query: String,
    schema: SchemaRef,
}

impl FlightSqlTableProvider {
    /// Connect to `endpoint`, execute `query` once to retrieve the Arrow
    /// schema and return a provider that can be registered into a
    /// [`SessionContext`](datafusion::prelude::SessionContext).
    pub async fn try_new(
        endpoint: impl Into<String>,
        query: impl Into<String>,
    ) -> Result<Self, FlightProviderError> {
        let endpoint = endpoint.into();
        let query = query.into();

        let mut client = connect(&endpoint).await?;
        let info = client
            .execute(query.clone(), None)
            .await
            .map_err(|e| FlightProviderError::Flight(e.to_string()))?;
        let schema = info
            .try_decode_schema()
            .map_err(|e| FlightProviderError::Flight(e.to_string()))?;

        Ok(Self {
            endpoint,
            query,
            schema: Arc::new(schema),
        })
    }

    /// Build a provider with a known schema, skipping the upfront network
    /// round-trip used by [`Self::try_new`]. Useful when the schema is part
    /// of a service contract or has already been negotiated out-of-band.
    ///
    /// * `endpoint` – a `tonic`-compatible URI (e.g. `http://host:port`).
    /// * `query`    – the SQL statement to issue against the remote service.
    /// * `schema`   – the Arrow schema of the result-set.
    pub fn new(
        endpoint: impl Into<String>,
        query: impl Into<String>,
        schema: SchemaRef,
    ) -> Self {
        Self {
            endpoint: endpoint.into(),
            query: query.into(),
            schema,
        }
    }

    /// Returns the configured Flight SQL endpoint URL.
    /// The remote endpoint URI this provider talks to.
    pub fn endpoint(&self) -> &str {
        &self.endpoint
    }

    /// Returns the SQL statement that will be sent to the remote endpoint.
    pub fn query(&self) -> &str {
        &self.query
    }
}

#[async_trait]
impl TableProvider for FlightSqlTableProvider {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn schema(&self) -> SchemaRef {
        Arc::clone(&self.schema)
    }

    fn table_type(&self) -> TableType {
        TableType::Base
    }

    async fn scan(
        &self,
        _state: &dyn Session,
        projection: Option<&Vec<usize>>,
        _filters: &[Expr],
        limit: Option<usize>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        let exec = FlightSqlExec::try_new(
            self.endpoint.clone(),
            self.query.clone(),
            Arc::clone(&self.schema),
            projection.cloned(),
            limit,
        )?;
        Ok(Arc::new(exec))
    }

    /// Filter push-down is intentionally not implemented: the caller is
    /// expected to bake any predicates into the SQL `query` they hand to
    /// [`FlightSqlTableProvider::new`]. We report `Unsupported` so DataFusion
    /// re-evaluates filters locally on the batches we return.
    fn supports_filters_pushdown(
        &self,
        filters: &[&Expr],
    ) -> DfResult<Vec<TableProviderFilterPushDown>> {
        Ok(vec![TableProviderFilterPushDown::Unsupported; filters.len()])
    }
}

/// [`ExecutionPlan`] that streams batches out of a remote Flight SQL endpoint.
///
/// The plan owns the connection parameters and produces a single, bounded
/// partition. Projection and `LIMIT` are applied locally on the incoming
/// stream because the Flight SQL spec doesn't define a portable way to
/// rewrite the original statement in the remote dialect.
pub(crate) struct FlightSqlExec {
    endpoint: String,
    query: String,
    /// Schema visible to the rest of the plan after projection.
    projected_schema: SchemaRef,
    projection: Option<Vec<usize>>,
    limit: Option<usize>,
    properties: PlanProperties,
}

impl FlightSqlExec {
    fn try_new(
        endpoint: String,
        query: String,
        schema: SchemaRef,
        projection: Option<Vec<usize>>,
        limit: Option<usize>,
    ) -> DfResult<Self> {
        let projected_schema = project_schema(&schema, projection.as_ref())?;
        let properties = PlanProperties::new(
            EquivalenceProperties::new(Arc::clone(&projected_schema)),
            Partitioning::UnknownPartitioning(1),
            EmissionType::Incremental,
            Boundedness::Bounded,
        );
        Ok(Self {
            endpoint,
            query,
            projected_schema,
            projection,
            limit,
            properties,
        })
    }
}

impl fmt::Debug for FlightSqlExec {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("FlightSqlExec")
            .field("endpoint", &self.endpoint)
            .field("query", &self.query)
            .field("projection", &self.projection)
            .field("limit", &self.limit)
            .finish()
    }
}

impl DisplayAs for FlightSqlExec {
    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "FlightSqlExec: endpoint={}, query={:?}, projection={:?}, limit={:?}",
            self.endpoint, self.query, self.projection, self.limit
        )
    }
}

impl ExecutionPlan for FlightSqlExec {
    fn name(&self) -> &str {
        "FlightSqlExec"
    }

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn properties(&self) -> &PlanProperties {
        &self.properties
    }

    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
        vec![]
    }

    fn with_new_children(
        self: Arc<Self>,
        _children: Vec<Arc<dyn ExecutionPlan>>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        Ok(self)
    }

    fn execute(
        &self,
        _partition: usize,
        _context: Arc<TaskContext>,
    ) -> DfResult<SendableRecordBatchStream> {
        let endpoint = self.endpoint.clone();
        let query = self.query.clone();
        let projection = self.projection.clone();
        let limit = self.limit;
        let projected_schema = Arc::clone(&self.projected_schema);
        let projected_schema_for_stream = Arc::clone(&projected_schema);

        // The Flight SQL client API is `async`, but `ExecutionPlan::execute`
        // is synchronous. We materialise the connect / execute / do_get
        // dance lazily inside an async stream so DataFusion can drive it on
        // its own runtime.
        let batch_stream = stream::once(async move {
            run_remote_query(endpoint, query)
                .await
                .map_err(DataFusionError::from)
        })
        .try_flatten()
        .map(move |batch_res| {
            batch_res.and_then(|batch| project_batch(&batch, projection.as_deref()))
        });

        let limited = apply_limit(batch_stream, limit);

        Ok(Box::pin(RecordBatchStreamAdapter::new(
            projected_schema_for_stream,
            limited,
        )))
    }
}

/// Connects to a remote Flight SQL endpoint and returns a ready-to-use client.
async fn connect(endpoint: &str) -> Result<FlightSqlServiceClient<Channel>, FlightProviderError> {
    let ep = Endpoint::from_shared(endpoint.to_owned()).map_err(|source| {
        FlightProviderError::InvalidEndpoint {
            endpoint: endpoint.to_owned(),
            source,
        }
    })?;
    let channel = ep
        .connect()
        .await
        .map_err(|source| FlightProviderError::Connect {
            endpoint: endpoint.to_owned(),
            source,
        })?;
    Ok(FlightSqlServiceClient::new(channel))
}

/// Opens a fresh Flight SQL connection, executes `query` and returns a
/// `RecordBatch` stream that yields the union of every endpoint reported by
/// the remote server.
async fn run_remote_query(
    endpoint: String,
    query: String,
) -> Result<
    impl futures::Stream<Item = DfResult<RecordBatch>> + Send + 'static,
    FlightProviderError,
> {
    let mut client = connect(&endpoint).await?;
    let info = client
        .execute(query, None)
        .await
        .map_err(|e| FlightProviderError::Flight(e.to_string()))?;

    if info.endpoint.is_empty() {
        return Err(FlightProviderError::MissingTicket);
    }

    // Walk every endpoint's ticket and concatenate the resulting batch
    // streams. For most servers there will only be one endpoint.
    let mut streams = Vec::with_capacity(info.endpoint.len());
    for ep in info.endpoint {
        let ticket = ep.ticket.ok_or(FlightProviderError::MissingTicket)?;
        let stream = client
            .do_get(ticket)
            .await
            .map_err(|e| FlightProviderError::Flight(e.to_string()))?;
        streams.push(stream);
    }

    let merged = stream::iter(streams)
        .flatten()
        .map(|batch_res| batch_res.map_err(|e| DataFusionError::External(Box::new(e))));

    Ok(merged)
}

/// Apply the requested projection (a column-index list) to a batch.
fn project_batch(batch: &RecordBatch, projection: Option<&[usize]>) -> DfResult<RecordBatch> {
    match projection {
        Some(cols) => batch
            .project(cols)
            .map_err(|e| DataFusionError::ArrowError(e, None)),
        None => Ok(batch.clone()),
    }
}

/// Wrap the underlying batch stream so it stops once `limit` rows have been
/// emitted, slicing the boundary batch as needed.
fn apply_limit<S>(
    stream: S,
    limit: Option<usize>,
) -> impl futures::Stream<Item = DfResult<RecordBatch>> + Send
where
    S: futures::Stream<Item = DfResult<RecordBatch>> + Send + 'static,
{
    let enforce = limit.is_some();
    let budget = limit.unwrap_or(usize::MAX);
    futures::stream::unfold(
        (stream.boxed(), budget, enforce),
        |(mut s, mut budget, enforce)| async move {
            if enforce && budget == 0 {
                return None;
            }
            match s.next().await {
                None => None,
                Some(Err(e)) => Some((Err(e), (s, budget, enforce))),
                Some(Ok(batch)) => {
                    if !enforce {
                        return Some((Ok(batch), (s, budget, enforce)));
                    }
                    let rows = batch.num_rows();
                    if rows <= budget {
                        budget -= rows;
                        Some((Ok(batch), (s, budget, enforce)))
                    } else {
                        let sliced = batch.slice(0, budget);
                        Some((Ok(sliced), (s, 0, enforce)))
                    }
                }
            }
        },
    )
}

File: openfoundry-go/libs/query-engine/flight_provider.go
// Apache Arrow Flight SQL provider for OpenFoundry.
//
// FlightSQLProvider turns a remote Flight SQL endpoint into a
// callable scan. Any data-plane service can therefore federate
// queries against another Flight SQL service exactly as if its
// result-set were a local table:
//
//	provider, err := queryengine.TryNewFlightSQLProvider(ctx,
//	    "http://flight-sql.internal:50051",
//	    "SELECT id, name FROM customers")
//	if err != nil { return err }
//	rdr, err := provider.Scan(ctx, []int{0, 1}, 10)
//	if err != nil { return err }
//	defer rdr.Release()
//	for rdr.Next() { rec := rdr.RecordBatch(); … rec.Release() }
//	if err := rdr.Err(); err != nil { return err }
//
// # Push-down support
//
//   - Projection is pushed down trivially: only the requested columns
//     are kept once the batches arrive from the remote endpoint.
//   - Limit is pushed down trivially: the local stream stops as soon
//     as `limit` rows have been emitted, slicing the boundary batch
//     as needed.
//   - Filter push-down is intentionally out of scope. Translating
//     arbitrary expression trees into the SQL dialect of an unknown
//     remote engine cannot be done safely in a generic provider.
//     Callers that need filter push-down should embed the predicate
//     directly in the `query` they pass to TryNewFlightSQLProvider.
//
// # Divergence from the Rust crate
//
// The Rust crate exposes `FlightSqlTableProvider` as a DataFusion
// `TableProvider` plus an `ExecutionPlan` (FlightSqlExec). Go has no
// DataFusion equivalent, so we expose the same underlying capability
// — connect, execute, walk endpoints, project, limit — as a plain
// type with a Scan method that returns an `array.RecordReader`. The
// wire-level behaviour, error taxonomy, and push-down policy are
// identical to the Rust port.
package queryengine

import (
	"context"
	"errors"
	"fmt"
	"io"
	"strings"

	"github.com/apache/arrow-go/v18/arrow"
	"github.com/apache/arrow-go/v18/arrow/array"
	"github.com/apache/arrow-go/v18/arrow/flight"
	"github.com/apache/arrow-go/v18/arrow/flight/flightsql"
	"github.com/apache/arrow-go/v18/arrow/memory"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
)

// ─── Errors ────────────────────────────────────────────────────────────

// FlightProviderErrorKind tags FlightProviderError variants. Mirrors
// the Rust `enum FlightProviderError`.
type FlightProviderErrorKind uint8

const (
	// FlightInvalidEndpoint — the supplied endpoint URL could not be
	// parsed or reached.
	FlightInvalidEndpoint FlightProviderErrorKind = iota
	// FlightConnect — the TCP/HTTP-2 connection to the remote Flight
	// server failed.
	FlightConnect
	// FlightServer — the remote Flight SQL server returned an
	// Arrow-level error.
	FlightServer
	// FlightMissingTicket — the Flight server replied with a
	// FlightInfo that did not contain any endpoint (and therefore no
	// ticket) for the query.
	FlightMissingTicket
)

// FlightProviderError is the typed error returned by every
// FlightSQLProvider call. Use errors.As to discriminate.
type FlightProviderError struct {
	Kind     FlightProviderErrorKind
	Endpoint string
	Message  string
	Cause    error
}

func (e *FlightProviderError) Error() string {
	switch e.Kind {
	case FlightInvalidEndpoint:
		return fmt.Sprintf("invalid Flight SQL endpoint %q: %s", e.Endpoint, e.causeOrMsg())
	case FlightConnect:
		return fmt.Sprintf("failed to connect to Flight SQL endpoint %q: %s", e.Endpoint, e.causeOrMsg())
	case FlightServer:
		return "Flight SQL error: " + e.causeOrMsg()
	case FlightMissingTicket:
		return "Flight SQL endpoint returned no ticket for query"
	default:
		return "flight provider error: " + e.causeOrMsg()
	}
}

func (e *FlightProviderError) Unwrap() error { return e.Cause }

func (e *FlightProviderError) causeOrMsg() string {
	if e.Message != "" {
		return e.Message
	}
	if e.Cause != nil {
		return e.Cause.Error()
	}
	return "unknown"
}

// IsInvalidEndpoint reports whether err is a FlightInvalidEndpoint.
func IsInvalidEndpoint(err error) bool { return matchFlightKind(err, FlightInvalidEndpoint) }

// IsConnectError reports whether err is a FlightConnect.
func IsConnectError(err error) bool { return matchFlightKind(err, FlightConnect) }

// IsServerError reports whether err is a FlightServer.
func IsServerError(err error) bool { return matchFlightKind(err, FlightServer) }

// IsMissingTicket reports whether err is a FlightMissingTicket.
func IsMissingTicket(err error) bool { return matchFlightKind(err, FlightMissingTicket) }

func matchFlightKind(err error, kind FlightProviderErrorKind) bool {
	var fe *FlightProviderError
	return errors.As(err, &fe) && fe.Kind == kind
}

// ─── Provider ──────────────────────────────────────────────────────────

// FlightSQLProvider resolves its data by issuing a SQL query against
// a remote Apache Arrow Flight SQL endpoint. Mirrors the Rust
// FlightSqlTableProvider.
//
// The schema is supplied up-front (either by the caller via
// NewFlightSQLProvider or by an upfront execute round-trip via
// TryNewFlightSQLProvider) because callers typically need it before
// any I/O takes place. The remote service must produce a result-set
// that is compatible with this schema.
type FlightSQLProvider struct {
	endpoint string
	query    string
	schema   *arrow.Schema
	dialOpts []grpc.DialOption
	alloc    memory.Allocator
}

// FlightProviderOption customises a FlightSQLProvider at construction
// time.
type FlightProviderOption func(*flightProviderOpts)

type flightProviderOpts struct {
	dialOpts []grpc.DialOption
	alloc    memory.Allocator
}

// WithDialOptions overrides the gRPC dial options used to connect
// to the remote Flight SQL endpoint. When unset the provider infers
// transport credentials from the URL scheme: `http://` and bare
// `host:port` use insecure credentials; `https://` uses system TLS.
func WithDialOptions(opts ...grpc.DialOption) FlightProviderOption {
	return func(o *flightProviderOpts) { o.dialOpts = opts }
}

// WithAllocator overrides the Arrow allocator used to materialise
// projected batches. Defaults to memory.DefaultAllocator.
func WithAllocator(alloc memory.Allocator) FlightProviderOption {
	return func(o *flightProviderOpts) { o.alloc = alloc }
}

// TryNewFlightSQLProvider connects to `endpoint`, executes `query`
// once to retrieve the Arrow schema, and returns a provider ready to
// be scanned. Mirrors the Rust try_new constructor.
func TryNewFlightSQLProvider(ctx context.Context, endpoint, query string, opts ...FlightProviderOption) (*FlightSQLProvider, error) {
	options := resolveOpts(endpoint, opts)
	client, err := dialFlight(ctx, endpoint, options.dialOpts)
	if err != nil {
		return nil, err
	}
	defer client.Close()
	info, err := client.Execute(ctx, query)
	if err != nil {
		return nil, &FlightProviderError{Kind: FlightServer, Cause: err}
	}
	schema, err := flight.DeserializeSchema(info.GetSchema(), options.alloc)
	if err != nil {
		return nil, &FlightProviderError{Kind: FlightServer, Cause: err}
	}
	return &FlightSQLProvider{
		endpoint: endpoint,
		query:    query,
		schema:   schema,
		dialOpts: options.dialOpts,
		alloc:    options.alloc,
	}, nil
}

// NewFlightSQLProvider builds a provider with a known schema, skipping
// the upfront network round-trip used by TryNewFlightSQLProvider.
// Useful when the schema is part of a service contract or has already
// been negotiated out-of-band. Mirrors the Rust new constructor.
func NewFlightSQLProvider(endpoint, query string, schema *arrow.Schema, opts ...FlightProviderOption) *FlightSQLProvider {
	options := resolveOpts(endpoint, opts)
	return &FlightSQLProvider{
		endpoint: endpoint,
		query:    query,
		schema:   schema,
		dialOpts: options.dialOpts,
		alloc:    options.alloc,
	}
}

// Endpoint returns the configured Flight SQL endpoint URL.
func (p *FlightSQLProvider) Endpoint() string { return p.endpoint }

// Query returns the SQL statement that will be sent to the remote
// endpoint.
func (p *FlightSQLProvider) Query() string { return p.query }

// Schema returns the result-set schema reported by the remote
// endpoint (or supplied to NewFlightSQLProvider).
func (p *FlightSQLProvider) Schema() *arrow.Schema { return p.schema }

// Scan opens a fresh connection, executes the query, walks every
// endpoint's ticket and returns a RecordReader that streams the
// concatenated batches with the requested projection and limit
// applied locally.
//
// `projection` is a list of column indices (nil = all columns).
// `limit` caps the total number of rows emitted (negative or zero
// = unbounded), slicing the boundary batch as needed.
//
// The caller must Release the returned reader. Mirrors
// `TableProvider::scan` + `FlightSqlExec::execute` in the Rust port.
func (p *FlightSQLProvider) Scan(ctx context.Context, projection []int, limit int) (array.RecordReader, error) {
	projectedSchema, err := projectSchema(p.schema, projection)
	if err != nil {
		return nil, err
	}
	client, err := dialFlight(ctx, p.endpoint, p.dialOpts)
	if err != nil {
		return nil, err
	}
	info, err := client.Execute(ctx, p.query)
	if err != nil {
		client.Close()
		return nil, &FlightProviderError{Kind: FlightServer, Cause: err}
	}
	if len(info.GetEndpoint()) == 0 {
		client.Close()
		return nil, &FlightProviderError{Kind: FlightMissingTicket}
	}
	tickets := make([]*flight.Ticket, 0, len(info.GetEndpoint()))
	for _, ep := range info.GetEndpoint() {
		tk := ep.GetTicket()
		if tk == nil {
			client.Close()
			return nil, &FlightProviderError{Kind: FlightMissingTicket}
		}
		tickets = append(tickets, tk)
	}
	return &flightStreamReader{
		ctx:        ctx,
		client:     client,
		tickets:    tickets,
		projection: append([]int(nil), projection...),
		limit:      limit,
		schema:     projectedSchema,
		alloc:      p.alloc,
	}, nil
}

// ─── Stream reader ─────────────────────────────────────────────────────

// flightStreamReader implements array.RecordReader by walking the
// list of Flight tickets sequentially, applying the requested
// projection + limit on every batch as it arrives. Mirrors the Rust
// `FlightSqlExec::execute` flow (`stream::iter(streams).flatten()` +
// `apply_limit`).
type flightStreamReader struct {
	ctx        context.Context
	client     *flightsql.Client
	tickets    []*flight.Ticket
	tIdx       int
	current    *flight.Reader
	projection []int
	schema     *arrow.Schema
	alloc      memory.Allocator

	limit       int
	enforce     bool
	budget      int
	budgetReady bool

	last    arrow.RecordBatch
	err     error
	closed  bool
	refCnt  int64
	started bool
}

// Schema returns the projected schema. Implements array.RecordReader.
func (r *flightStreamReader) Schema() *arrow.Schema { return r.schema }

// Retain bumps the reference count. Implements array.RecordReader.
func (r *flightStreamReader) Retain() { r.refCnt++ }

// Release drops the reference count and closes the underlying client
// when the count reaches zero. Implements array.RecordReader.
func (r *flightStreamReader) Release() {
	if r.refCnt > 0 {
		r.refCnt--
		if r.refCnt > 0 {
			return
		}
	}
	r.close()
}

// Err returns the terminal error, if any. Implements
// array.RecordReader.
func (r *flightStreamReader) Err() error { return r.err }

// RecordBatch returns the most recently read batch. Implements
// array.RecordReader.
func (r *flightStreamReader) RecordBatch() arrow.RecordBatch { return r.last }

// Record returns the most recently read batch. Deprecated alias for
// RecordBatch retained for backward compatibility with the
// array.RecordReader interface.
func (r *flightStreamReader) Record() arrow.RecordBatch { return r.last }

// Next advances to the next batch. Returns false at end-of-stream
// or on error (use Err to discriminate).
func (r *flightStreamReader) Next() bool {
	if r.closed || r.err != nil {
		return false
	}
	if !r.budgetReady {
		r.enforce = r.limit > 0
		r.budget = r.limit
		r.budgetReady = true
	}
	if r.enforce && r.budget <= 0 {
		return false
	}
	if r.last != nil {
		r.last.Release()
		r.last = nil
	}
	for {
		batch, ok := r.advance()
		if !ok {
			return false
		}
		projected, err := projectBatch(r.alloc, batch, r.projection)
		batch.Release()
		if err != nil {
			r.err = err
			return false
		}
		if r.enforce {
			rows := int(projected.NumRows())
			if rows == 0 {
				projected.Release()
				continue
			}
			if rows > r.budget {
				sliced := projected.NewSlice(0, int64(r.budget))
				projected.Release()
				projected = sliced
				r.budget = 0
			} else {
				r.budget -= rows
			}
		}
		r.last = projected
		return true
	}
}

// advance pulls the next raw batch from the underlying flight
// readers, opening fresh ones as tickets are consumed.
func (r *flightStreamReader) advance() (arrow.RecordBatch, bool) {
	for {
		if r.current == nil {
			if r.tIdx >= len(r.tickets) {
				return nil, false
			}
			rdr, err := r.client.DoGet(r.ctx, r.tickets[r.tIdx])
			if err != nil {
				r.err = &FlightProviderError{Kind: FlightServer, Cause: err}
				return nil, false
			}
			r.tIdx++
			r.current = rdr
			r.started = true
		}
		if r.current.Next() {
			rec := r.current.RecordBatch()
			rec.Retain()
			return rec, true
		}
		if cerr := r.current.Err(); cerr != nil && !errors.Is(cerr, io.EOF) {
			r.err = &FlightProviderError{Kind: FlightServer, Cause: cerr}
			r.current.Release()
			r.current = nil
			return nil, false
		}
		r.current.Release()
		r.current = nil
	}
}

func (r *flightStreamReader) close() {
	if r.closed {
		return
	}
	r.closed = true
	if r.last != nil {
		r.last.Release()
		r.last = nil
	}
	if r.current != nil {
		r.current.Release()
		r.current = nil
	}
	if r.client != nil {
		_ = r.client.Close()
		r.client = nil
	}
}

// ─── Helpers ───────────────────────────────────────────────────────────

// dialFlight is the Go analogue of `connect` in the Rust port.
// Strips the URL scheme so gRPC sees a `host:port` target, then
// dials the remote endpoint with caller-supplied (or scheme-inferred)
// credentials.
func dialFlight(ctx context.Context, endpoint string, dialOpts []grpc.DialOption) (*flightsql.Client, error) {
	target := normaliseFlightEndpoint(endpoint)
	if target == "" {
		return nil, &FlightProviderError{Kind: FlightInvalidEndpoint, Endpoint: endpoint, Message: "empty endpoint"}
	}
	if dialOpts == nil {
		dialOpts = []grpc.DialOption{inferTransportCredentials(endpoint)}
	}
	client, err := flightsql.NewClientCtx(ctx, target, nil, nil, dialOpts...)
	if err != nil {
		return nil, &FlightProviderError{Kind: FlightConnect, Endpoint: endpoint, Cause: err}
	}
	return client, nil
}

// normaliseFlightEndpoint strips well-known schemes + trailing slash
// so the gRPC dialer sees a `host:port` target. Mirrors the same
// helper in services/sql-bi-gateway-service/internal/flightsql.
func normaliseFlightEndpoint(endpoint string) string {
	t := strings.TrimSpace(endpoint)
	for _, prefix := range []string{"http://", "https://", "grpc://", "grpc+tcp://", "grpc+tls://"} {
		t = strings.TrimPrefix(t, prefix)
	}
	return strings.TrimRight(t, "/")
}

// inferTransportCredentials picks insecure credentials for plain
// `http://` (and bare `host:port`) and system TLS for `https://` /
// `grpc+tls://`.
func inferTransportCredentials(endpoint string) grpc.DialOption {
	low := strings.ToLower(strings.TrimSpace(endpoint))
	if strings.HasPrefix(low, "https://") || strings.HasPrefix(low, "grpc+tls://") {
		return grpc.WithTransportCredentials(credentials.NewTLS(nil))
	}
	return grpc.WithTransportCredentials(insecure.NewCredentials())
}

// projectSchema returns the projection of `schema` keeping only the
// columns whose indices appear in `cols`. Passing nil returns the
// schema unchanged. Mirrors `datafusion::common::project_schema`.
func projectSchema(schema *arrow.Schema, cols []int) (*arrow.Schema, error) {
	if cols == nil {
		return schema, nil
	}
	fields := make([]arrow.Field, len(cols))
	for i, c := range cols {
		if c < 0 || c >= len(schema.Fields()) {
			return nil, &FlightProviderError{
				Kind:    FlightServer,
				Message: fmt.Sprintf("projection index %d out of range for schema with %d columns", c, len(schema.Fields())),
			}
		}
		fields[i] = schema.Field(c)
	}
	md := schema.Metadata()
	return arrow.NewSchema(fields, &md), nil
}

// projectBatch slices a batch down to the requested column indices.
// Returns a freshly-retained batch the caller must Release. Passing
// a nil projection returns the batch unchanged (with an extra Retain
// so the caller may always Release).
func projectBatch(alloc memory.Allocator, batch arrow.RecordBatch, cols []int) (arrow.RecordBatch, error) {
	if cols == nil {
		batch.Retain()
		return batch, nil
	}
	projectedSchema, err := projectSchema(batch.Schema(), cols)
	if err != nil {
		return nil, err
	}
	arrays := make([]arrow.Array, len(cols))
	for i, c := range cols {
		if c < 0 || c >= int(batch.NumCols()) {
			return nil, &FlightProviderError{
				Kind:    FlightServer,
				Message: fmt.Sprintf("projection index %d out of range for batch with %d columns", c, batch.NumCols()),
			}
		}
		arrays[i] = batch.Column(c)
		arrays[i].Retain()
	}
	rec := array.NewRecordBatch(projectedSchema, arrays, batch.NumRows())
	for _, a := range arrays {
		a.Release()
	}
	_ = alloc // reserved for future builder-based projections
	return rec, nil
}

// resolveOpts merges caller-supplied options with the defaults.
func resolveOpts(endpoint string, opts []FlightProviderOption) flightProviderOpts {
	resolved := flightProviderOpts{}
	for _, o := range opts {
		o(&resolved)
	}
	if resolved.alloc == nil {
		resolved.alloc = memory.DefaultAllocator
	}
	if resolved.dialOpts == nil {
		resolved.dialOpts = []grpc.DialOption{inferTransportCredentials(endpoint)}
	}
	return resolved
}
